crates/kitsune2/tests/blocks.rs

```diff
@@ -0,0 +1,498 @@
+use kitsune2_test_utils::iter_check;
+use kitsune2_test_utils::{
+    agent::TestLocalAgent, enable_tracing, space::TEST_SPACE_ID,
+};
+use kitsune2_transport_tx5::harness::Tx5TransportTestHarness;
+use std::sync::mpsc::{Receiver, Sender};
+
+/// A default mock space handler that just drops received messages.
+#[derive(Debug)]
+struct MockSpaceHandler {
+    recv_notify_sender: Sender<bytes::Bytes>,
+}
+
+impl MockSpaceHandler {
+    fn create() -> (Self, Receiver<bytes::Bytes>) {
+        let (send, recv) = std::sync::mpsc::channel();
+        (
+            Self {
+                recv_notify_sender: send,
+            },
+            recv,
+        )
+    }
+}
+
+impl SpaceHandler for MockSpaceHandler {
+    fn recv_notify(
+        &self,
+        _from_peer: Url,
+        _space_id: kitsune2_api::SpaceId,
+        data: bytes::Bytes,
+    ) -> K2Result<()> {
+        self.recv_notify_sender.send(data).unwrap();
+        Ok(())
+    }
+}
+
+/// A mock TxHandler
+#[derive(Debug)]
+pub struct MockTxHandler {
+    pub peer_url: std::sync::Mutex<Url>,
+    space: Arc<Mutex<OnceCell<DynSpace>>>,
+    recv_module_msg_sender: Sender<bytes::Bytes>,
+}
+
+impl MockTxHandler {
+    fn create(
+        peer_url: Mutex<Url>,
+        space: Arc<Mutex<OnceCell<DynSpace>>>,
+    ) -> (Arc<Self>, Receiver<bytes::Bytes>) {
+        let (recv_module_msg_sender, recv) = std::sync::mpsc::channel();
+
+        (
+            Arc::new(Self {
+                peer_url,
+                space,
+                recv_module_msg_sender,
+            }),
+            recv,
+        )
+    }
+}
+
+impl TxModuleHandler for MockTxHandler {
+    fn recv_module_msg(
+        &self,
+        _peer: Url,
+        _space_id: kitsune2_api::SpaceId,
+        _module: String,
+        data: bytes::Bytes,
+    ) -> K2Result<()> {
+        self.recv_module_msg_sender.send(data).unwrap();
+        Ok(())
+    }
+}
+
+impl TxBaseHandler for MockTxHandler {
+    fn new_listening_address(&self, this_url: Url) -> BoxFut<'static, ()> {
+        *(self.peer_url.lock().unwrap()) = this_url;
+        Box::pin(async {})
+    }
+}
+impl TxHandler for MockTxHandler {
+    fn preflight_gather_outgoing(
+        &self,
+        _peer_url: Url,
+    ) -> BoxFut<'_, K2Result<bytes::Bytes>> {
+        let space = self.space.lock().expect("poisoned");
+        let space = space
+            .get()
+            .expect("Space OnceCell has not been initialized in time.")
+            .clone();
+
+        Box::pin(async move {
+            let agents = space.peer_store().get_all().await.unwrap();
+            let agents_encoded: Vec<String> =
+                agents.into_iter().map(|a| a.encode().unwrap()).collect();
+            let mut b = bytes::BufMut::writer(bytes::BytesMut::new());
+            rmp_serde::encode::write_named(&mut b, &agents_encoded).unwrap();
+            Ok(b.into_inner().freeze())
+        })
+    }
+
+    fn preflight_validate_incoming(
```

Comment:
(Partially) simulates the behavior that we have in holochain where agents received via preflight messages are added to the peer store.